site/en/external/lockfile.md

```diff
@@ -224,3 +224,47 @@ practices:
 By following these best practices, you can effectively utilize the lockfile
 feature in Bazel, leading to more efficient, reliable, and collaborative
 software development workflows.
+
+## Merge Conflicts {:#merge-conflicts}
+
+The lockfile format is designed to minimize merge conflicts, but they can still
+happen.
+
+### Automatic Resolution {:#automatic-resolution}
+
+Bazel provides a custom
+[git merge driver](https://git-scm.com/docs/gitattributes#_defining_a_custom_merge_driver)
```

Comment:
git merge drivers are clunky because of the setup steps.

I assume you considered having the same semantics as tools like `npm` which include the understanding of merge conflicts in their parser/resolver of the lockfile? Such that no external tooling would be needed?

Reply:
I agree that the one-time setup is clunky, but the usability after that is actually better than what `npm` offers: With `npm`, you have to resolve the conflict in `package.json` first and then run `npm install`. With this merge driver, you don't even get a conflict in `MODULE.bazel.lock` reported in the first place, regardless of what the conflict state of `MODULE.bazel` is.

We could theoretically build functionality into `bazel` that resolves the merge conflict in `MODULE.bazel.lock` even if `MODULE.bazel` contains an unresolved conflict, but that would either require a dedicated command or accepting that whatever Bazel command you run for this fail, but still updates the lockfile. That doesn't sound very discoverable.

Since only the initial setup is clunky, I hope that the contextual error message pointing to it when a merge conflict is observed will help make this usable overall.

What do you think about the tradeoff?

Reply:
Sorry I missed the reply.

> whatever Bazel command you run for this fail, but still updates the lockfile

I think ideally a `bazel build` with conflicts in the lockfile would update the lockfile (maybe interactive prompt) and then succeed. Perhaps that can be built into one of the wrapper CLIs that already has interactivity.

Anyhow I really only meant to ask if you considered it, and not to block this being merged. Glad to see it's landed now!